### ISSUE TITLE
[RFC] ./configure checks failing for 2nd order dependencies during build due to environment/linking failures

### ISSUE THREAD

**Issue by gfodor** [habitat-sh/habitat#3303]
I have found that when building custom packages, `configure` can fail to detect the presence of pulled in dependent libraries, if those libraries are part of a 2nd or higher order dependency. There are two root causes of this.

For the sake of the example, assume I have a package `foo/C` which I am building which depends upon `foo/B` (which is published) which in turn depends upon `foo/A` (also published.)

When building `foo/C`, the generated `configure` script (included in the source tarball) may check for capabilities present in libraries from either `foo/B` or `foo/A` that rely upon the ability to compile and link against `foo/A`'s headers and libraries. At build time of `foo/C`, habitat does correctly download and install `foo/B` and `foo/A` recursively into the local `/hab/pkgs` directory, but `configure` checks can erroneously fail for two reasons.

First, the library + pkg config paths for `foo/A` are not properly included into the build environment by `hab-plan-build` before `configure` is run. So, any `pkg-config` based checks against `foo/A`'s libraries will fail, and `gcc` will also not be passed the library paths via `-L` flags. Depending on the check, `configure` can rely upon both of these things. In practice, including `foo/A` explicitly in the `pkg_deps` of `foo/C` can work around this problem, but it seems logical that at build time `foo/C` should incorporate the library/pkg-config paths into environment variables not just of `foo/B` but also of its dependencies, recursively.

Second, often times a `configure` check will be checking for dynamic linking of a library via `-lXXX` where `XXX` is a library of a dependency. For example, if the package `foo/C` has `configure` set up to assert that `foo/B` links via `-l`, this requires that `ld` has the ability to *also* load `foo/A`'s libraries. The problem is that even if we include `foo/A` in `pkg_deps` (which will push its library paths to `gcc` via `-L`), these checks can still fail since `ld` will still fail to actually load the shared libraries from `foo/A` unless:

- The library paths are passed in to `ld` via the `gcc` switch `-Wl,-rpath-link=<path>`, or
- The paths to `foo/A`'s libraries are included on `LD_LIBRARY_PATH` or `LIBRARY_PATH`

Note that including `foo/A`'s library paths via the `-L` flag to `gcc` does *not* seem to be sufficient to ensure that `ld` will load the libraries for `foo/A` for a `configure` check, unless those library names are passed in explicitly via the `-l` switch of `gcc` in addition to those of `foo/B`. In practice, `configure` scripts generated by `autoconf` do not do this.

At build time, habitat writes out `LD_RUN_PATH` to include library paths to all of the direct runtime dependencies of `foo/C`, namely those of `foo/B` (but not `foo/A`.)

For `configure` to pass all checks given these issues, the workaround I've devised for now to build a package `foo/C` is:

- Include `foo/A` in `pkg_deps` directly, even though it is already in the `pkg_deps` of `foo/B`, so `hab-plan-build` will include its library paths into `LD_RUN_PATH` and `LDFLAGS`, properly add it to `PKG_CONFIG_PATH`, and add its include paths to `CPPFLAGS`.
- Update my `plan.sh` to set `LD_LIBRARY_PATH=$LD_RUN_PATH` in `do_build()` to ensure that `ld` will be able to load `foo/A`'s libraries during `configure`.

Of course, unless I am missing something, all of the above implies that there may be a large number of `configure` checks silently failing if they are optional checks. Required checks hard fail, but optional checks may be failing on 2nd+ order dependencies unless they are explicitly pulled in via the techniques above.

In other words, if a library `X` depends on `Y` depends on `Z`, there may be published packages for `X` that silently have the features disabled that rely upon `Y`, if those features are optional and the `configure` script for `X` disables them by trying to build against `Z`'s headers or tries to dynamically load `Y`'s libraries that are dependent on `Z`. 

Any help on this is appreciated!

**Comment by gfodor** [comment]
As one concrete example, any `configure` check that currently checks against `librt` (part of `core/glibc`) will fail with a default `plan.sh` if there are symbols pointing to `libpthread` in the stub that `configure` generates. For example, `configure` checking for `clock_gettime` will fail on a default `plan.sh` that includes `core/glib` as a 'pkg_dep' right now since `gcc` fails when `ld` tries to load `libpthread`, and `gcc` is invoked with just `gcc -lrt`. (`librt` depends on `libpthread`)

(the workaround, as mentioned, is to update `LD_LIBRARY_PATH` to `LD_RUN_PATH` in the `plan.sh`)

**Comment by eeyun** [comment]
@fnichol do you have any feedback on this just yet? Reading through this it seems pretty close in nature to  #2658 ? I might be mistaken on that. We'll wait to sort this into the backlog until we can get some comments here.

**Comment by rsertelon** [comment]
This seems to me that the reasoning behind current behavior of habitat is sound and should be kept that way.

Maybe it is `./configure` that has a limitation to work in this isolated way? And if it is the case, the workaround shown by @gfodor might be documented somewhere?

**Comment by gfodor** [comment]
I don't think the issue has to do with configure, fundamentally any build
process that relies upon linking 2nd order dependencies could be affected
if they rely upon the linker the way I outlined in my post, configure just
incidentally seems to be an easy way to reproduce the problems

**Comment by rsertelon** [comment]
When you say `linking 2nd order dependencies` do you mean linking currently built binary to it, or linking 1st order dep to 2nd?

**Comment by gfodor** [comment]
I probably can't describe it better than my write up, I context switched
off this stuff a while ago so it's not as fresh in my mind

**Comment by rsertelon** [comment]
Ok ok it's just that I'm a Java developer. Had some C courses but long ago :) thus the questions ;)

**Comment by rsertelon** [comment]
I've been creating the gtk2 package recently, and I think I now understand better what you meant @gfodor It seems that gtk builds have added a "2nd order dependency check" which helps a bit more but does not "solve the problem" you wrote about.

Judging by the experience I had, I understand your point correctly I think. However, I still think that the decision of the core team to prevent linking with 2nd order dependencies is a good thing.

This way, all runtime dependencies of the package that is to be built/installed are described explicitly. Being explicit is important IMHO as we make sure that a rebuild of bldr will not break the app if one of our dependencies decide (for example), to remove one of theirs, that we needed.

This is the responsibility of the package maintainer, IMHO, to make sure that all links are satisfied for every binaries and libraries created for the package. We may need tools to be sure we are doing the right thing more easily than running ldd by hand on every binaries/libraries :)

**Comment by gfodor** [comment]
I think if you have a dependency chain of packages where X depends on Y depends on Z, then if `configure` for X is checking for APIs in Z, then it makes sense for the package maintainer to have to add Z to the plan.

However, the thing I've highlighted is broken *regardless* if X depends on Z, at all. Linking against Y fails too! In general, `ld` will fail to link Y against X if it is not able to find Z, and it won't be able to unless a) Z is included in the plan for X so its paths are added to the environment and b) the path to Z is also manually added to `LD_LIBRARY_PATH`.

**Comment by gfodor** [comment]
Also FWIW I think the correct solution to this problem is to make it so at build time, `LD_LIBRARY_PATH` (and perhaps `LD_RUN_PATH`) includes the entire dependency tree's library paths, so that dynamic linking actually works properly for any libraries that have their own dependencies.

**Comment by adamhjk** [comment]
I really go back and forth on this. On the one hand, I absolutely see how much easier life would be if we pulled all the transitive dependencies in this way (in particular, it means you don't have to go spelunking around in the specifics of how a given configure script works.) 

On the other, there is a very real risk of behavior being introduced through a transitive change in a dependency. For example, package A has behavior that can be turned on if `libz` is available. The plan for A does not want `libz` enabled. But it does depend on package B; and package B uses `libz`. If we enact the above, then a change in plan B would introduce, silently, a new run-time behavior in A. The next time we build A, we would automatically, without consent, enable the behavior of B, unless it was explicitly turned off.

Now, the user experience question is really pointed at by the last sentence in the above paragraph. We could decide that if you *really* want control over having a given feature absolutely, positively disabled, we want you to pass `--without` to your `configure` script. This would probably mean that the system DWIM's more often than not ("I just want to build package B, because I need it for A, but I have to do so much jamming of shit into B"). I do think it will then come as a gotcha to one of our foundational principles (that you get the same build every time, because you have to be explicit about what goes in the bag.) 

@ryankeairns @eeyun @fnichol what do you think?

**Comment by eeyun** [comment]
I think we've finally come to a consensus on the team today in triage and we are in agreement that (at least for now) we would not want to accept a change of this nature. The general feeling is that trading off some up-front pain for potential extreme pain later in an app's lifecycle isn't one we are currently willing to make.

That being said. We are always willing to re-evaluate decisions we're making in the project so it is possible that if we have more users who feel like this is absolutely something they need in order to be productive we will most definitely re-evaluate the decision. For now though we're going to close this one out. Thank you to everyone who shared their thoughts on the subject and thank you for the original submissing @gfodor. We're going to be adding a follow up issue that references this one to add more clarity to our documentation around WHY users must be explicit in their dependency declarations.
